hazelcast/src/test/java/com/hazelcast/map/IndexStatsChangingNumberOfMembersTest.java

```diff
@@ -0,0 +1,395 @@
+        assertEquals(originalMap1AverageHitLatency, valueStats(map1).getAverageHitLatency());
+        assertEquals(originalMap1AverageHitSelectivity, valueStats(map1).getAverageHitSelectivity(), 0.001);
+        assertEquals(originalMap3QueryCount, stats(map3).getQueryCount());
+        assertEquals(originalMap3IndexedQueryCount, stats(map3).getIndexedQueryCount());
+        assertEquals(originalMap3IndexQueryCount, valueStats(map3).getQueryCount());
+        assertEquals(originalMap3AverageHitLatency, valueStats(map3).getAverageHitLatency());
+        assertEquals(originalMap3AverageHitSelectivity, valueStats(map3).getAverageHitSelectivity(), 0.001);
+
+        assertEquals(originalOverallAverageHitSelectivity,
+                calculateOverallSelectivity(map2Hits, map2TotalHitSelectivity, map1, map3), 0.001);
+
+        for (int i = 0; i < queriesBulk; i++) {
+            map3.entrySet(Predicates.alwaysTrue());
+            map1.entrySet(Predicates.equal("this", 10));
+            map3.entrySet(Predicates.lessEqual("this", lessEqualCount));
+        }
+
+        assertEquals(9 * queriesBulk, stats(map1).getQueryCount());
+        assertEquals(6 * queriesBulk, stats(map1).getIndexedQueryCount());
+        assertEquals(6 * queriesBulk, valueStats(map1).getQueryCount());
+        assertEquals(6 * queriesBulk, stats(map3).getQueryCount());
+        assertEquals(4 * queriesBulk, stats(map3).getIndexedQueryCount());
+        assertEquals(4 * queriesBulk, valueStats(map3).getQueryCount());
+
+        // This work correctly only due to we stored data from shutdown member and uses this data for counting
+        // originalOverallAverageHitSelectivity. However this not represent real scenario. This check is here just for ensure
+        // that AverageHitSelectivity is still counted correctly on live members.
+        originalOverallAverageHitSelectivity = calculateOverallSelectivity(map2Hits, map2TotalHitSelectivity, map1, map3);
+        assertEquals((expectedEqual + expectedGreaterEqual) / 2, originalOverallAverageHitSelectivity, 0.015);
+    }
+
+    @Test
+    public void testIndexStatsOperationChangingNumberOfMembers() {
+        int inserts = 100;
+        int updates = 20;
+        int removes = 20;
+
+        String mapName = randomMapName();
+        Config config = getConfig();
+        config.getMapConfig(mapName).setInMemoryFormat(inMemoryFormat);
+
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(NODE_COUNT);
+        HazelcastInstance instance1 = factory.newHazelcastInstance(config);
+        HazelcastInstance instance2 = factory.newHazelcastInstance(config);
+        IMap<Integer, Integer> map1 = instance1.getMap(mapName);
+        IMap<Integer, Integer> map2 = instance2.getMap(mapName);
+
+        map1.addIndex("this", false);
+        map2.addIndex("this", false);
+
+        assertEquals(0, valueStats(map1).getInsertCount());
```

Comment:
I would use a local variable `localIndexStats1 = valueStats(map1)` to speed up the test. The same for `map2`.
Or is there any reason to construct the new `LocalIndexStats` in each call?

There are also other places within the tests in this PR which seem to have the same behavior.

Reply:
It's for a reason: some parts of map stats, including index stats, are recomputed on-demand, to update those parts `getLocalMapStats` should be called. Seems like that is not documented, I will update `IMap#getLocalMapStats` javadocs.